src/main.cairo

```diff
@@ -1,32 +1,231 @@
+    let (erc20, renewal_price) = Pricing.compute_renew_price(pricing, root_domain, 365);
+
+    let (contract) = get_contract_address();
+    with_attr error_message("Error transferring tokens from renewer to contract") {
+        IERC20.transferFrom(erc20, renewer, contract, renewal_price);
+    }
+
+    Naming.renew(naming, root_domain, 365);
+    
+    domain_renewed.emit(root_domain, renewer, 365);
+
+    return ();
+}
+
+@external
+func batch_renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain_len: felt, domain: felt*, renewer_len: felt, renewer: felt*
+) {
+    with_attr error_message("domain and renewer array must have same length") {
+        assert domain_len = renewer_len;
+    }
+    return _batch_renew_iter(domain_len, domain, renewer_len, renewer);
+}
+
+func _batch_renew_iter{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain_len: felt, domain: felt*, renewer_len: felt, renewer: felt*
+) {
+    if (domain_len == 0) {
+        return ();
+    }
+    renew(domain[0], renewer[0]);
+    return _batch_renew_iter(domain_len - 1, domain + 1, renewer_len - 1, renewer + 1);
+}
+
+// @notice Vote for an upgrade
+// @param upgrade Upgrade to vote for
+@external
+func vote_upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    upgrade_id: felt, implementation_hash: felt, vote: felt
+) {
+    let (caller) = get_caller_address();
+
+    with_attr error_message("Votes can either be 1 or 0") {
+        assert vote * (vote - 1) = 0;
+    }
+
+    voted_upgrade.write(caller, upgrade_id, implementation_hash, vote);
+    voted.emit(caller, upgrade_id, implementation_hash, vote);
     return ();
 }
+
+// @notice Upgrade implementation
+// @param new_implementation hash
+@external
+func upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    upgrade_id: felt, implementation_hash: felt,
+) {
+    with_attr error_message("Upgrade is blacklisted") {
+        let (is_blacklisted) = blacklisted.read(upgrade_id);
+        assert is_blacklisted = 0;
+    }
+    let (vote_sum) = _upgrade_iter(upgrade_id, implementation_hash, VOTERS_LEN, 0);
+    with_attr error_message("Not enough votes to upgrade") {
+        assert_le(VOTING_QUORUM, vote_sum);
+    }
+
+    Proxy._set_implementation_hash(implementation_hash);
+    blacklisted.write(upgrade_id, 1);
+    return ();
+}
+
+func _upgrade_iter{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    upgrade_id: felt,
+    implementation_hash: felt,
+    voters: felt,
+    vote_sum: felt,
+) -> (vote_sum: felt) {
+    if (voters == 0) {
+        return (vote_sum,);
+    }
+    let (voters_data_start_label) = get_label_location(voters_data_start);
+    let voters_arr = cast(voters_data_start_label, felt*);
+    let (vote) = voted_upgrade.read(voters_arr[voters - 1], upgrade_id, implementation_hash);
+    return _upgrade_iter(upgrade_id, implementation_hash, voters - 1, vote_sum + vote);
```

Comment:
Why do we need the upgrade ID and not only the new implementation hash ?

Reply:
We need a couple upgrade_id and implementation_hash to ensure that an implementation that was already passed cannot be passed again. Also in case we want to roll back to a previous implementation, if the implementation_hash is already blacklisted it won't be possible.